### ISSUE TITLE
Train with fit_generator fails while regular fit works, please help!

### ISSUE THREAD

**Issue by vicmak** [fchollet/keras#4407]
Hi !
I am trying to train a network on a large dataset (Microsoft sentence completion challenge) and I have to split the train data to batches since it does not fit into the GPU memory. Oddly enough preparing the train in the same exact manner, but with a gnerator that yields the X and Y - I receive the following error:

> Exception: Error when checking model input: expected dense_input_1 to have shape (None, 1200) but got array with shape (1200L, 1L)

each observation instance is actually an array of 1200 floats.
Yes, removed the dimension before yield, since it is one observation and not a set. 
the model is defined as following:

    word_dimension = 300
    context_size = 4
    hidden_layer_size = 100
    vocabulary_size = 30000
    log_file_name = "C:\\corpora\\MSCC\\log_mscc.txt"
    model_save_file_name = "C:\\corpora\\models\\model.h5"
    start_time = time.time()
    print("Creating the model object")
    model = Sequential()
    model.add(Dense(hidden_layer_size, input_dim=context_size * word_dimension, init='uniform', activation='tanh'))
    model.add(Dense(vocabulary_size, init='normal', activation='softmax'))  # can be also sigmoid (for a multiclass)
    print("compiling...")
    model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
    print("compiled!")

the code is [here](https://github.com/vicmak/ProofSeer/blob/master/CorpusLM2.py), and the difference in functions train_model_from_dir_batch and train_model_from_dir

The train call differs in the following way:

on complete fit:

    model.fit(arrX, arrY, nb_epoch=50, batch_size=dataset.shape[0])

on fit_generator:

    model.fit_generator(n_gram_generator(vocabulary,current_filename, vectors), samples_per_epoch=100, nb_epoch=50)

Please make sure that the boxes below are checked before you submit your issue. Thank you!

- [V] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [V] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [V] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

Thanks a lot!

**Comment by carlthome** [comment]
Seems your generator is yielding transposed ndarrays.

**Comment by vicmak** [comment]
thanks for the fast reply!
That is what I thought of first.
But the thing is that the generator yields 1-dimensional array. how can it be transposed? anyway I tried to call np.transpose(X) and still get the same exception(and the same shape when of X)

If the dataset in single fit operation is a 2 dimensional array, where every row is a sample from train dataset, then generator's yield should return 1-dimensional array representing a single sample. The number of samples in the complete batch is then specified in samples_per_epoch parameter.

Do I understand this correct?

Thanks!

**Comment by keunwoochoi** [comment]
> The number of samples in the complete batch is then specified in samples_per_epoch parameter.
Probably not, it's `samples_per_epoch`, not `samples_per_batch`.

When I used `fit_generator()`, my generator yielded (batch_size, blah blah) shaped data. In other words, it was a **batch generator** and worked fine. So `samples_per_batch` was set to be `N * batch_size` for my convenience.

**Comment by carlthome** [comment]
Yeah, @keunwoochoi is right. If you're yielding one dimensional arrays you're using `fit_generator` wrong. The generator should yield batches. This is all in the documentation.

**Comment by vicmak** [comment]
thank you guys so much @carlthome @keunwoochoi !
You are right.
By the way, I finally use a workaround of a single for loop calling fit method, with yield generator of my desired batch size.

Thanks again!

**Comment by flinn2112** [comment]
Even this post is from 2016, I ran into the same issue, and biting on it for hours.
carlthome's answer is the key but needs to be clarified:
"..you're using fit_generator wrong. The generator should yield batches. ..."


1.  a generator is just a different way to feed the model fit args:
